Adds JSON/CBOR support and an Io-type option

cheqianh opened the pull request:
# Description: 
This PR integrates JSON/CBOR into the benchmark CLI. It additionally adds an IO type option to decide which IO type of the Ion data will be benchmarked. 

# Details

In details, this PR adds below features/components
1. Enable the CLI tool to benchmark Ion data as well as modern JSON/CBOR libraries concurrently.
2. Adds an IO type option to allow the tool benchmark time for either read/write in-memory buffers or files.
3. A bunch of refactor and unit testing.

# Output Example
## Example One 
Benchmark load API for all JSON libraries as well as ION

**Command and Output:**
```
python amazon/ionbenchmark/ion_benchmark_cli.py read --format json --format orjson --format simplejson --format ion_text --format ion_binary --format ujson
```
<img width="800" alt="Screenshot 2023-03-09 at 5 53 25 PM" src="https://user-images.githubusercontent.com/67451029/224203064-ca80d9a6-9fb6-44ef-b789-24044acc4be4.png">


## Example Two 
Benchmark dumps API (--io-types buffer) for CBOR2,  and Ion_binary

**Command and Output:**
```
python amazon/ionbenchmark/ion_benchmark_cli.py write --io-type buffer --format cbor2 --format ion_binary test_unit_int
```
<img width="808" alt="Screenshot 2023-03-09 at 5 56 19 PM" src="https://user-images.githubusercontent.com/67451029/224203459-049ac635-3621-4129-92fd-1db8aca949d1.png">


## Example Three  
Benchmark dump (--io-types file) API for CBOR2,  and Ion_binary

**Command and Output:**
```
python amazon/ionbenchmark/ion_benchmark_cli.py write --io-type file --format cbor2 --format ion_binary test_unit_int
```
<img width="800" alt="Screenshot 2023-03-09 at 5 56 58 PM" src="https://user-images.githubusercontent.com/67451029/224203540-a1df9daf-d4f8-4fbd-877c-ddccf03c6c89.png">



# Follow-up issues
1. **Memory profiling enhancement**  
Currently, the tool only profile a memory usage peak. Due to the potential optimization Python may do, the memory usage metrics might be inaccurate after running different formats multiple execution times. Need more investigation for memory related metrics - https://github.com/amazon-ion/ion-python/issues/245
2. **Needs a tool to convert different file formats.** 
https://github.com/amazon-ion/ion-python/issues/234
3. **Needs pretty printed format options**
Right now, the options output in the output table is unclear. For example, (simpleion, ion_binary, file) represents that `--api` is simpleion, `--format` is ion_binary and `--io-type` is file. This needs improvement - https://github.com/amazon-ion/ion-python/issues/244. 





By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

cheqianh commented:
The CI/CD failed because pypy does not support CBOR2.

cheqianh reviewed (commented):
The CI/CD failed because of that `pypy` does not support `cbor2`. We need to either,
(1) Disable all cbor2 feature for `pypy`, or
(2) treat `cbor2` as `cbor` for `pypy` environment

cheqianh reviewed (commented):
CI/CD failed due some json libraries does not support `pypy`. E.g., orjson

Without that, it works as expected with known 10 failed tests - https://github.com/amazon-ion/ion-python/issues/249

linlin-s commented on tests/test_benchmark_cli.py:
```diff
@@ -135,39 +136,87 @@ def gather_all_options_in_list(table):
+@parametrize(
+    *tuple((f.value for f in Format.Format if Format.format_is_json(f.value)))
+)
+def test_write_json_format(f):
```
Should we also include the test case for `test_write_cbor_format` here?

cheqianh replied:
[87e7f65](https://github.com/amazon-ion/ion-python/pull/243/commits/87e7f65f9d05fc527820226a5dc93c9ee1d22954) addressed them.

tgregg commented:
> CI/CD failed due some json libraries does not support `pypy`. E.g., orjson. We can either
> 
>     1. Treat `orjson` as regular built-in json library when using `pypy` interpreter, or
> 
>     2. Remove orjson support on pypy.
> 
> 
> Without that, it works as expected with known 10 failed tests - #249

Let's go with option 2

tgregg commented on ion-c:
Since this introduces test failures that haven't yet been resolved, let's leave the ion-c submodule update for a separate PR, and revert it here.

tgregg commented on amazon/ionbenchmark/ion_benchmark_cli.py:
For the `--api` option, the allowed values should be changed to something like `load_dump` and `streaming`. The option `simple_ion` is no longer accurate for the other formats, and looks weird in the options list when those formats are selected.

cheqianh replied:
Changed.

tgregg commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -84,52 +100,169 @@
+
+
+# Generates benchmark code for json/cbor load/loads APIs
+def generate_read_test_code(file, memory_profiling, format_option, binary, io_type):
```
Is there a reason we can't follow the same path for Ion?

cheqianh replied:
Merged them into one.

tgregg commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -264,33 +431,126 @@ def write_micro_benchmark_simpleion(iterations, warmups, c_extension, obj, file,
+
+
+# Generates benchmark code for json dump API
+def generate_write_test_code(obj, memory_profiling, format_option, io_type, binary):
```
Is there a reason Ion can't follow this path?

cheqianh replied:
same as above comment.

tgregg commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -360,55 +646,50 @@ def ion_python_benchmark_cli(arguments):
+        # reset each option configuration
+        api, format_option, io_type = reset_for_each_execution(each_option)
+        binary = format_is_binary(format_option)
+        # TODO. currently, we must provide the tool a corresponding file format for read benchmarking. For example,
```
The tool *to convert to* a corresponding file format for read benchmarking. ?

cheqianh replied:
Fixed.

tgregg commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -84,52 +100,169 @@
-                    it = ion.load(fp, single_value=single_value, emit_bare_values=emit_bare_values,
-                                  parse_eagerly=False)
+            if not iterator:
+                def test_func():
+                    tracemalloc.start()
+                    data = ion.loads(benchmark_data, single_value=single_value, emit_bare_values=emit_bare_values)
+                    global read_memory_usage_peak
+                    read_memory_usage_peak = tracemalloc.get_traced_memory()[1] / BYTES_TO_MB
+                    tracemalloc.stop()
+                    return data
```
`test_func()` is what is being timed, correct? That means the cost of using `tracemalloc` is included in the results. Is there a way for us to extract this logic outside of the timed block?

cheqianh replied:
Yes, we execute it twice; once with `memory_profiling` enabled, and once with it disabled. 

We profile the memory before all performance benchmarking [here](https://github.com/amazon-ion/ion-python/blob/master/amazon/ionbenchmark/ion_benchmark_cli.py#L284-L287). Then we [benchmark](https://github.com/amazon-ion/ion-python/blob/master/amazon/ionbenchmark/ion_benchmark_cli.py#L289-L293) the performance separately.

tgregg reviewed (commented):
Since format conversion is left as a TODO, can you explain what is actually happening today in each of your examples in the description? Do those come from an actual execution?

Once we have data conversion, I'd expect the data size stat to be different for each of the formats, as it will reflect the size of the *converted* data, for both read and write.

cheqianh commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -55,24 +55,40 @@
+     -i --io-type <io_type>             The source or destination type, from the set (buffer | file). If buffer is
+                                        selected, buffers the input data in memory before reading and writes the output
+                                        data to an in-memory buffer instead of a file. [default: file]
+
```
Thank you catching this, added.

cheqianh commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -318,8 +578,7 @@ def insert_into_report_table(table, row):
         return identify_report_table_helper(
-            ['file_size (MB)', 'options', 'total_time (s)', 'conversion_\ntime (s)', 'conversion_time/\ntotal_time (%)',
-             'memory_usage_peak (MB)'])
+            ['file_size (MB)', 'options', 'total_time (s)', 'memory_usage_peak (MB)'])
```
Yep, we should unify the read/write tables ultimately. I left it separately for now because Read/Write command may generate different metrics. 

One example is the previous `READ` table, which includes an additional `conversion time` column to indicate the execution time of whether raw Python objects are wrapped into simpleIon types.

linlin-s replied:
Thanks for explaining.

cheqianh commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -84,52 +100,169 @@
+                def test_func():
+                    with open(file, "br") as fp:
+                        data = ion.load(fp, single_value=single_value, emit_bare_values=emit_bare_values)
+                    return data
```
> if that's possible to merge the similar logic in two different conditions into one reusable function?

Good question. We may not want to merge them into one function now but it's possible to refactor some logics if `test_func()` becomes complex later.

This is because the whole functionality of the `test_func` will be benchmarked and they (`load` and `loads`) are different APIs with a similar signature.

cheqianh replied:
> if that's possible to merge the similar logic in two different conditions into one reusable function?

Good question. We may not want to merge them into one function now but it's possible to refactor some logics if `test_func()` becomes complex later.

This is because the whole functionality of the `test_func` will be benchmarked and they (`load` and `loads`) are different APIs with a similar signature.

cheqianh commented:
> Since format conversion is left as a TODO, can you explain what is actually happening today in each of your examples in the description? Do those come from an actual execution?

Yep, these metrics are generated by the tool directly. Overall, the tool is assuming that we had the format conversion functionality already but it doesn't do any conversion work for files. 

For each example,
(1) Benchmarking both read and write for JSON/Ion_text doesn't require format conversion. One thing I want to point out is that ion_binary is compared but it's doing the exact same thing as format `ion_text`. We had an enhancement ticket describing this - https://github.com/amazon-ion/ion-python/issues/234. Once we supported format conversion, the tool will update it automatically without any changes. 

(2) and (3)
Benchmarking `write` command for CBOR/Ion doesn't require format conversion; we just need to provide python layer objects and doesn't need to care about how they encode the objects and write them into files. But I found one thing that breaks the fairness of Ion/CBOR benchamrking is the slight difference in python objects generated by them (E.g., cbor2 reads "9233720363654371807" as `-12852` in python). But the tool will automatically generate the correct object once we support the format conversion. (This is the tenet when I work on this PR; I always assume that we had the formats conversion tool already so `cbor2` will generate correct `cbor2` objects from the desired file format)

> Once we have data conversion, I'd expect the data size stat to be different for each of the formats, as it will reflect the size of the converted data, for both read and write.

Yep, the [converted file](https://github.com/amazon-ion/ion-python/pull/243/files#diff-193f9a476a85a550a2aa80e3a9ad54272c2661544bc96455fd6439664e1b9df1R666) is being used for calculating the file size. Once we support the formats conversion, it will generate new file size automatically.

tgregg commented on amazon/ionbenchmark/ion_benchmark_cli.py:
```diff
@@ -94,6 +94,9 @@
 if not pypy:
     import tracemalloc
+    import orjson
+else:
+    import json as orjson
```
I think we should just throw if `orjson` is requested when using pypy, rather than silently replacing it.

cheqianh replied:
Yeah, I try to push other comments first to see if it pass the CI/CD, still working on it.

cheqianh commented:
There are few `pypy` issue failing the CI/CD, working on it.

cheqianh commented:
The new commit addressed all the feedback above, passed all the CI/CD and `pypy` incompatible issue, and deprecated `orjson` for now. 

Here is the GH [issue](https://github.com/amazon-ion/ion-python/issues/253) to add orjson back.
